Upgrade Jackson to 2.5.1

ctrimble opened the pull request:
This PR is not complete.

Changes required to support Jackson 2.5.1.  Currently, this is just the POM updates required, exposing one failing test related to `MinKey` and `MaxKey` fields.  Other than that issue, this project seems compatible with Jackson 2.5.1.

ctrimble commented:
I am still seeing this error in the test suite:

```
  canHandleMinAndMaxKey(org.jongo.marshall.DocumentMarshallingTest): Unable to unmarshall result to class org.jongo.marshall.DocumentMarshallingTest$BSONPrimitiveType from content { "_id" : { "$oid" : "5632afd0ae4fc18d6822877b"} , "minKey" : { "$minKey" : 1} , "maxKey" : { "$maxKey" : 1}}
```

ctrimble commented:
@bguerout do you have any clue why min and max key would be breaking after this upgrade?

ctrimble commented:
The issue with @JsonProperty overrides on the `_id` field seems to be unrelated to the upgrade and should be in a separate PR.  The override can be accomplished with mixin annotations like this:

```
@MongoId
@MongoObjectId
@JsonProperty("_id")
```

ctrimble commented:
The way that the compatibility test suite works is not friendly to mixIn annotations.  Some ideas:

- Supply two different annotation introspectors, one for the legacy stuff and one for the new annotations
- Just drop the old annotations

ctrimble commented:
The way that \@MongoId and \@MongoObjectId are handled in this project may not be the best.  Jackson's [BeanSerializationModifier](https://fasterxml.github.io/jackson-databind/javadoc/2.6/com/fasterxml/jackson/databind/ser/BeanSerializerModifier.html) and [BeanDeserializationModifier](https://fasterxml.github.io/jackson-databind/javadoc/2.6/com/fasterxml/jackson/databind/deser/BeanDeserializerModifier.html) are better for tasks like this.

bguerout commented:
Thanks for the pull request. Sorry but I haven't found the time to look at it. I'm going to read to tomorrow.

ctrimble commented:
@bguerout this PR needs some cleanup and should probably be a couple of PRs.  There are several things happening:

1. Moving to Jackson 2.5.1 is breaking serialization of min key and max key fields
2. A change was introduced in Jongo (version 1.2?) that broke Jongo's annotations when used with mixins.  I added a parameterized test to pin down the functionality.
3. I added JUnit test rules for embedded Mongo and Jongo, to support using the parameterized test runner.

ctrimble commented:
Created #274 to deal with mix in annotations not working.  Rebased this work on that PR.

ctrimble commented:
Looks like there is an outstanding test for MinKey in bson4jackson https://github.com/michel-kraemer/bson4jackson/pull/50

bguerout commented on src/main/java/org/jongo/marshall/jackson/JacksonMapper.java:
```diff
@@ -73,7 +72,7 @@ public Mapper build() {
                 queryFactory = new BsonQueryFactory(jacksonEngine);
             }
             if (objectIdUpdater == null) {
-                objectIdUpdater = new ReflectiveObjectIdUpdater(new JacksonIdFieldSelector());
```
Well done. This is this end of this too long living life class

ctrimble replied:
Should the classes that are no longer being used be totally removed from the project, or are deprecation warnings the way you want to go?

bguerout commented on src/main/java/org/jongo/marshall/jackson/JongoAnnotationIntrospector.java:
```diff
@@ -16,19 +16,50 @@
+        return hasMongoObjectId ? Include.NON_NULL : defValue;
     }
 
+    @Override
```
hasMongoObjectId and hasMongoId can be extracted into private methods

ctrimble replied:
Is there already a good location for those in the project?

bguerout commented on src/test/java/org/jongo/model/IdSpecSet.java:
```diff
@@ -0,0 +1,143 @@
+    } 
+  }
+  
+  public static <T> T specInstance(Class<T> cls, ObjectId id) {
```
It's a bit hard to understand what this method does by reading its name.

ctrimble replied:
Agreed.  How about `idForSpec`?

ctrimble replied:
Oh, perhaps you need `specInstance`.  I can see that.  Perhaps `newInstanceForSpec` with the class argument named `spec` would clear things up.

bguerout commented:
First of all thanks I would like to thank for this pull request. You have refactored a lot of code. This is cool.
All tests are OK if Jackson and Bson4jackson versions are rolled back.
Are you OK to update this pull request with older version of theses two libraires to be able to merge it ?

ctrimble commented:
@bguerout take a look at #274.  That does not have the version upgrades in it.

ctrimble commented:
Rebased this PR on master, to reduce confusion.  All your notes were implemented.

bguerout commented:
Issue https://github.com/michel-kraemer/bson4jackson/issues/51 in bson4jackson project has been fixed.
This pull request will be merged when bson4jackson will be released.

ctrimble commented:
Totally awesome!  Thanks @bguerout !  Ping me if you need any action on my end.

bguerout commented:
BTW according to this comment https://github.com/michel-kraemer/bson4jackson/pull/52#issuecomment-154828045, the next version of bson4jackson will be 2.6

ctrimble commented:
I saw that.  I find it strange that he does not release a 2.5.1, since this is a patch, and then test against  2.6.  That would more naturally indicate the intended version matrix.
